Frontend location CRUD ops

benjenkinsv95 opened the pull request:
### Please give a short summary of what's included in this submission
Adding frontend pages for viewing all locations, creating, editing, and deleting them.

### Checklist
- [x] The Pull Request title describes the changes I've made
- [x] I've reviewed the **Files changed** tab
- [x] The code in the **Files changed** tab meets our style guides
- [x] New classes and public methods in the **Files changed** tab are documented
- [x] I've added two reviewers for this pull request


### Frontend Reviewer Checklist
##### I've reviewed the observe page, reports page, and any page directly affected by these changes
- [x] For Mobile (width: ~320px) [View](http://whatismyscreenresolution.net/multi-screen-test?site-url=http://localhost:4200/login&w=320&h=480)
- [x] For Tablet landscape (width: ~1024px) [View](http://whatismyscreenresolution.net/multi-screen-test?site-url=http://localhost:4200/login&w=1024&h=768)
- [x] For Desktop (width: 1920px) [View](http://whatismyscreenresolution.net/multi-screen-test?site-url=http://localhost:4200/login&w=1920&h=1080)
- [x] I've reviewed the **Files Changed** Tab and left constructive criticism (if applicable)

###### Optional
- [ ] For Tablet portrait (width: ~768px) [View](http://whatismyscreenresolution.net/multi-screen-test?site-url=http://localhost:4200/login&w=768&h=1024)

benjenkinsv95 commented:
@stbenjam Would the cancel button just redirect back to the page that lists all locations?

stbenjam commented:
>  Furthermore, there doesn't seem to be a prevention to deleting a parent that has children. This could be part of the confirmation. 

I think this should probably get fixed on the back end, and if a user tries to do that we just return an error that it still has children.  Alternatively ask for confirmation in the UI and have the backend delete the children automatically, I'm not sure they should stick around if the parent goes away.

WesleySkillern commented:
I like the newly seed data. Makes the Reports more interesting!

WesleySkillern commented:
Just a question, are we only doing one layer of nesting? I can't seem to get children on the 'Update Location' page?

<img width="1390" alt="screen shot 2018-04-27 at 11 37 36 am" src="https://user-images.githubusercontent.com/36180018/39371371-7f31a7a0-4a0f-11e8-9064-6aeb5b65e151.png">

WesleySkillern commented:
Updating Locations doesn't seem to work for me.

<img width="1390" alt="screen shot 2018-04-27 at 11 40 36 am" src="https://user-images.githubusercontent.com/36180018/39371491-dc1eaabc-4a0f-11e8-986f-c2200eb70f4c.png">

WesleySkillern commented on frontend/src/app/location-edit/location-edit.component.ts:
```diff
@@ -0,0 +1,150 @@
+
+
+    submit(name: string, address: string, zip: string, country: string, type: string) {
+        if (this.nameFormControl.invalid) {
```
Nice assertions! These will never be triggered though, because the button will not be available until required fields are filled though.

benjenkinsv95 replied:
Those are mostly defensive checks. It prevents the really stupid malicious users who see a disabled button, open up chromes editor, and then enable it just to see what the fuck happens if they submit data anyway.

But I'm sure more intelligent malicious users could get around that easily, since it's still on the frontend. Can remove if you want.

benjenkinsv95 commented:
# Changes
Added a Cancel button.
![](https://i.imgur.com/3wUTm8R.png)

Added delete prompt that informs you that a location has child locations.
! [](https://i.imgur.com/anLcK2U.png)

# Answers to comments
> Any time making changes to the backend data, we should probably have a confirmation message. E.g. 'Are you sure?' Since it has the potential to change the app for all users. - @WesleySkillern 

I disagree. It interrupts the typical flow and an accidental change can be easily reverted in most cases. 

Ex. Editing a location's data (name, address, etc.) is easy to fix. But deleting a location causes someone to recreate the location and all of the links, so deleting should have a prompt.

I think Introducing confirmation messages should be thought through on a case by case basis. They don't make sense everywhere.

> I've noticed a couple issues with delete so far. If I click on the trash can for an item and refresh the page, the data just reappears. - @WesleySkillern 
I don't have that issue. Can you retry it now that Michael's changes have been merged in? My changes depended on some of his changes, but I wasn't pushing them myself with this PR. 

> Just a question, are we only doing one layer of nesting? I can't seem to get children on the 'Update Location' page. How did you create 'Room 1' or 'Room 2' located inside a Ward? - @WesleySkillern 
Multiple layers of nesting works fine for me. I think this might come back to not having Michael's fixes merged in yet.
![](https://i.imgur.com/P5NbKQ4.png)

> Updating Locations doesn't seem to work for me. - @WesleySkillern 
Can update them fine on my end.
![](https://i.imgur.com/MvSQ8D0.png)
Might be the the same dependency issue.

> I am able to create duplicate Name-Type combinations. Shouldn't Locations be unique by Name and Type since these are the only two required fields? - @WesleySkillern 
I'd request that as a backend change if it is easy enough. (@stbenjam)

> I'm also getting lots of 404's in the console on the page. - @WesleySkillern 
If a location doesn't have a parent it returns an error. So those are kind of expected? Maybe we could ask the backend for a better solution? (@stbenjam)

benjenkinsv95 commented:
@stbenjam I added a confirmation that that deleting the parent will delete the locations inside of it. Is that an easy enough change to make on the backend?

stbenjam commented:
@benjenkinsv95 I'm not sure how much work it is.  I would imagine Spring can cascade the delete or children locations, but time is rather short.  If it's easy enough to change the prompt, I would take that approach and give me a low priority card to take a look at fixing it?

benjenkinsv95 commented:
@stbenjam I'll change up the prompt then

WesleySkillern commented:
Review again. Nice work!
